fuzzers/fuzzer.ts

```diff
@@ -0,0 +1,386 @@
+	"(self::list-item) and parent::*[self::list]",
+	"fonto:document-type-node(.)",
+	"descendant-or-self::fn[not(parent::table-wrap-foot)]",
+	"ancestor-or-self::*[self::p][1]",
+	"(self::list-item) and parent::*[self::list[@list-type = 'roman-lower']]",
+	"(self::list-item) and parent::*[self::list[@list-type = 'arabic']]",
+	"(self::list-item) and parent::*[self::list[@list-type = 'alpha-lower']]",
+	"ancestor-or-self::*[fonto:is-table(.)]",
+	"self::xref[@ref-type='table-fn']",
+	"self::xref[@ref-type='bibr']",
+	"self::xref[@ref-type='fn']",
+	"ancestor::table[parent::table-wrap]",
+	"ancestor::*[not(fonto:block-layout(.))][1]",
+	"ancestor-or-self::term-sec",
+	"if($data('contextNode')[@list-type=('bullet', 'dash')]) then $data('contextNode')/@list-type else 'dash'",
+	"fonto:selection-common-ancestor()/ancestor-or-self::node()",
+	"ancestor-or-self::xref[@ref-type = 'fn']",
+	"if($data('contextNode')[@list-type=('alpha-lower', 'arabic', 'roman-lower')]) then $data('contextNode')/@list-type else 'alpha-lower'",
+	"ancestor-or-self::*[self::list and @list-type = 'alpha-lower'][1]",
+	"ancestor-or-self::*[self::list and @list-type = 'dash'][1]"
+];
+
+const uniqueStacks = new Set();
+
+let totalCases = 0;
+const startTime = process.hrtime();
+function worker(tid: number) {
+	while (true) {
+		// Record and print progress
+		if (tid === 0 && totalCases % 1000 === 0) {
+			const elapsed = process.hrtime(startTime);
+			const fcps = (totalCases / elapsed[0]).toFixed(2);
+			readline.clearLine(process.stdout, 0);
+			readline.cursorTo(process.stdout, 0, null);
+			process.stdout.write(
+				`\t[Total cases: ${totalCases}] [fcps: ${fcps}] [Unique crashes: ${uniqueStacks.size}]`
+			);
+		}
+		totalCases += 1;
+
+		// Select an expression from the corpus
+		let expression = corpus[rand(corpus.length)];
+
+		// Mutate the input using a simple character mutation
+		expression = mutateString(expression);
+		expression = mutateCharactersInPlace(expression);
+		// if (totalCases % 100 === 0) {
+		// 	process.stdout.write(`${expression}\n`);
+		// }
+
+		// Try to evaluate the expression
+		try {
+			evaluateXPath(expression, documentNode);
+		} catch (error) {
+			// Not interested in static errors, we're looking for crashes
+			if (error.message.startsWith('XPST')) {
+				continue;
+			}
+			// Not interested in type errors, we're looking for crashes
+			if (error.message.startsWith('XPTY')) {
+				continue;
+			}
+			// Not interested in type errors, we're looking for crashes
+			if (error.message.startsWith('FORG')) {
+				continue;
+			}
+
+			// Not interested in duplicate stack traces
+			if (uniqueStacks.has(error.stack)) {
+				continue;
+			}
+			uniqueStacks.add(error.stack);
+
+			// Print the error
+			process.stdout.write(
+				`\n\n!!! Found error !!\nInput: ${expression}\n${error.stack}\n\n`
+			);
+		}
+	}
+}
+
+// TODO: Figure out how to scale this to multiple cpu cores to fuzz faster
```

Comment:
`npm run fuzzer & npm run fuzzer & npm run fuzzer &`? :laughing:

Reply:
I have 4 cores+HT so your command would not utilize my CPU to it's fullest potential 🙃. It shouldn't be too hard (I hope) to get this done properly in nodejs.